### PR TITLE
8238274: (sctp) JDK-7118373 is not fixed for SctpChannel

### PR DESCRIPTION
**Pull request by masyano**
Please review this change to the Unix implementations of sun.nio.ch.sctp.Sctp*ChannelImpl#implCloseSelectableChannel() 
to be same as SocketChannelImpl at JDK-7118373. (The preClose is missing a check for the ST_KILLED state.)

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8238274](https://bugs.openjdk.java.net/browse/JDK-8238274): (sctp) JDK-7118373 is not fixed for SctpChannel

### Reviewers
 * [Daniel Fuchs](https://openjdk.java.net/census#dfuchs) (@dfuch - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/5274/head:pull/5274` \
`$ git checkout pull/5274`

Update a local copy of the PR: \
`$ git checkout pull/5274` \
`$ git pull https://git.openjdk.java.net/jdk pull/5274/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 5274`

View PR using the GUI difftool: \
`$ git pr show -t 5274`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/5274.diff">https://git.openjdk.java.net/jdk/pull/5274.diff</a>

</details>

### CODE REVIEW

**Comment by AlanBateman** [comment]
As the equivalent of JDK-7118373 in JDK 8 then the proposed change looks okay.

At some point we may have to look at aligning the SCTP channel implementations where close has been significantly re-implemented.

**Comment by dfuch** [comment]
I have run this change on one of our machines that support SCTP. I did get some intermittent failures with the other SCTP tests - they don't seem much stable - but the new proposed test was failing all the time. I suspect that using `lsof` to figure out whether the file descriptor was closed is not reliable/portable enough. I also tried to modify the test to use /othervm - which is probably a good idea if you don't want your results to be polluted by whatever other test might have run previously in the agent VM - but to no avail: the test was still failing 100% of the time.
@masyano could you figure another way to detect whether the file descriptor has been released?

**Comment by masyano** [comment]
@dfuch I tried to count `/proc/<PID>/fd`, but there are recorded all fds in the process. I don't know how to pick up only active fds in `/proc/<PID>/fd`. So, I will change to call lsof last one time only (like https://github.com/openjdk/jdk/pull/4621), and counts all unreleased fds.

Could you try this test case?

File: test/jdk/com/sun/nio/sctp/SctpChannel/CloseDescriptors.java [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+/*
+ * @test
+ * @bug 8238274
+ * @summary Potential leak file descriptor for SCTP
+ * @requires (os.family == "linux")
+ * @run main CloseDescriptors
```
Comment (dfuch) [inline test/jdk/com/sun/nio/sctp/SctpChannel/CloseDescriptors.java:29]: The test SctpMultiChannel/CloseDescriptors seems to work - at least it doesn't fail all the time - but it's using `@run main/othervm`. I believe you should make this test run in /othervm mode for more stability. I will test again with your later change and /othervm mode.

Comment (masyano) [reply test/jdk/com/sun/nio/sctp/SctpChannel/CloseDescriptors.java]: I pushed a fix to run the test in othervm mode. Please test again.

Comment (masyano) [reply test/jdk/com/sun/nio/sctp/SctpChannel/CloseDescriptors.java]: Thank you for your help. I applied your suggested fix and it works fine. But it takes about 4 minutes to be finished. So I added a timeout parameter to 250.

**Review by dfuch — APPROVED** [review]
LGTM

**Comment by dfuch** [comment]
@masyano Thanks for providing this fix. If you integrate as instructed in [1] I will sponsor this change for you.
 [1] https://github.com/openjdk/jdk/pull/5274#issuecomment-925674914

**Pull request closed without merging**